### ISSUE TITLE
Run X11 windows without windows manager decoration

### ISSUE THREAD

**Issue by davidhunter22** [microsoft/wslg#530]
Is it possible to run an X11 windows without the window manager decoration, I believe this is sometimes called frameless or borderless.

**Comment by davidhunter22** [comment]
I mean the buttons like the "x" that are common on most windows. On a native Linux box you can run windows without this decoration which is added by the window managed not the application. Attached is a screen shot of xeyes which has the decoration. 

![Screenshot 2021-10-21 103335](https://user-images.githubusercontent.com/8083830/138326384-b0bbd791-982f-48a9-91cf-e746de774997.png)
e

**Comment by davidhunter22** [comment]
Ah just found this, https://github.com/microsoft/wslg/discussions/67 which seems to answer my question on window managers

**Comment by AntonSmolkov** [comment]
This is “fullscreen” Rider IDE under WSLG. This box around also makes me feel sad☹️
![00C7925D-71E6-485A-B88E-BB50D100AF0E](https://user-images.githubusercontent.com/5318028/141668950-b0ec2552-63f0-40ba-898f-3991e0305099.jpeg)

**Comment by sweetwong** [comment]
![image](https://user-images.githubusercontent.com/39878472/147551907-33d0c57c-22cf-477f-b805-bce0a2e75acf.png)

Please add fullscreen feature
I don't want to see this white box

**Comment by GiGurra** [comment]
Single biggest show stopper for me for switching from `VcXsrv` -> `wslg`.
Second biggest show stopper is wslg windows not seemingly supporting windows tilling/snapping/windows+LR hotkeys

**Comment by hideyukn88** [comment]
@GiGurra, thanks for feedback, we are aware that the "second one" is asked various occasions.

Regarding to no window decorations, @AntonSmolkov @sweetwong and @GiGurra, do you want only when window is at fullscreen? and without those frame, how do you expect you can put back window to non-fullscreen? Or even no frame at windowed (non-fullscreen), how it's expected to be moved? Do know have any model scenario we can refer to? thanks!

**Comment by GiGurra** [comment]
No definitely not only full screen.

The best example is the IDE intellij idea. You can download the free community version to see what it currently looks like inside wslg, and compare with how it behaves with a native windows installation.

Basically it always runs without decoration (it renders its own menus and minimize/maximize buttons etc in its own format

**Comment by hideyukn88** [comment]
@GiGurra, thanks for info. 

By looking at native Linux case, particularly running on Ubuntu desktop, the window frame (and title bar) seems still rendered by their Window manager, but it has more consistent visuals with rest of the desktop. So, to clarify the request here would be, it is not removing the window decorations, but do have window decorations with match to the rest of desktop and themes. 

Regarding to fullscreen case, it looks on JetBrains IDE, "enter fullscreen mode" option (from View->Appearance) is not available on WSLg, but it is available on Ubuntu desktop. I will need to investigate why.

Also, once JetBrains completed the Wayland native support, then the current Window frame won't be used.

Thanks!

**Comment by GiGurra** [comment]
@hideyukn88 

Hh, yes you are right. I had completely forgotten that ubuntu does keep some decoration at the top (but not these thick lines all around). When I last used native ubuntu for my dev machine I was actually running some decoration removal thing, dont remember exactly what :)

**Comment by differentprogramming** [comment]
Can we at least change the color to make it less obtrusive on a dark themed screen?

**Comment by JotaRata** [comment]
TBH that would be a really cool feature

**Comment by zodiia** [comment]
Any update on this? WSL has gone out of preview and yet this issue is still present.

Most of third party softwares that emulates X Servers managed to use native windows. Yet WSLg still has this ugly window manager decoration (let's be honest for a second, it looks like it wasn't updated for at least 15 years), and it doesn't even support basic window snapping, even less the more advanced features that we have with a native window.

This is a huge deal breaker for WSLg. I was excited for the GUI apps support on Windows 10, but this alone makes me switch back to X410.

@hideyukn88 @Biswa96

**Comment by theonlyfoxy** [comment]
+1 for This Issue. VcXsrv is still better in Window Management than WSLg.

**Comment by hideyukn88** [comment]
@theonlyfoxy, thanks for checking, we are making progress. if you are on the latest WSLg and Windows 11 (all update applied), please create below file with below contents, then window can be snapped by Win+Left or Right key, thanks!

C:\\Users\\[Your User Name]\\.wslgconfig

```
[system-distro-env]
WSLG_USE_MSTSC=true
WESTON_RDP_WINDOW_SNAP_ARRANGE=true
WESTON_RDPRAIL_SHELL_LOCAL_MOVE=true
```

**Comment by theonlyfoxy** [comment]
> @theonlyfoxy, thanks for checking, we are making progress. if you are on the latest WSLg and Windows 11 (all update applied), please create below file with below contents, then window can be snapped by Win+Left or Right key, thanks!
> 
> C:\\Users\\[Your User Name]\\.wslgconfig
> 
> ```
> [system-distro-env]
> WSLG_USE_MSTSC=true
> WESTON_RDP_WINDOW_SNAP_ARRANGE=true
> WESTON_RDPRAIL_SHELL_LOCAL_MOVE=true
> ```
> 
>  

What about the borders around the window and style of buttons?
Wouldn't it be possible to integrate vcxsrv into WSLg and fallback X11 apps to it?

**Comment by cyraid** [comment]
Without borders, the window is offset from the top left. I've noticed this in my own development as well. Having a maximized borderless window results in:
![image](https://user-images.githubusercontent.com/7562753/225794822-b17a4bd5-8c1f-4eaf-9bac-f1627a483b91.png)

**Comment by zodiia** [comment]
@cyraid how did you manage to remove the window borders? I think it's the whole point of this issue and it seems you've found some kind of magic to achieve it :sweat_smile:

**Comment by cyraid** [comment]
@zodiia Oh, that's just a borderless window. You have to specify the `XInternAtom(d, "_MOTIF_WM_HINTS", False);` for xlib when creating the window (or `glfwWindowHint(GLFW_DECORATED, GLFW_FALSE);` for GLFW, etc.)

For programs that you haven't written, there's not much you can do other than ask Microsoft to include Decoration options with WSLG (or at least environment variables to disable decorations on windows) as they're the current window manager, unless you can use a custom window manager with WSLG which I'm not sure you can currently? (based off of what I've been reading)

But for those screenshots: With Chrome in Linux, you can set the window to use no titlebar (so I'm guessing they're just setting the decorator hints to false). Btw, to fix that Chrome issue, you can set the `Preferred Ozone platform` in `chrome://flags` to `Wayland` for a better experience.

**Comment by gaoyan1998** [comment]
Hi, is there a solution to this problem, now idea still has a big box wrapped around it

**Comment by zodiia** [comment]
> Hi, is there a solution to this problem, now idea still has a big box wrapped around it

No solution yet, however we might be getting closer to it with Jetbrains products specifically, as they have recently released features for integrating the top bar with the window, just like it is on windows. We might get something better for IntelliJ, CLion, ... soon

**Comment by andrescevp** [comment]
It will be nice if we could simply set the colors of the borders, I mean, there are some constraints totally understable regarding usability and compatibility, but, I believe the simples solution is the best one.

**Comment by sweetwong** [comment]
@zodiia I think window border color following Windows system theme(Light or Dark) is a simple solution.

**Comment by bitnom** [comment]
lol my linux broke and I wound up having to use WSL and now PyCharm ugly af. What are we doin here guys? Where's the code for these titlebars? We'll fix it ourselves.

**Comment by sweetwong** [comment]
Why hasn't a seemingly straightforward request been accommodated after several years? Alternatively, could you provide guidance on how we might resolve this issue ourselves?

**Comment by deadcombo** [comment]
ve found this can be overridden in Emacs with the `undecorated` frame parameter.

See https://www.gnu.org/software/emacs/manual/html_node/elisp/Management-Parameters.html#undecorated.
It mentions Motif Window Manager Hints: https://specifications.freedesktop.org/wm-spec/wm-spec-latest.html so maybe it can be generalized @hideyukn88 

Didn't manage to make it work in `early-init.el` but it does in `init.el`. Just `(push '(undecorated .t) default-frame-alist)`.

If you can't make it work for some reason, `toggle-frame-fullscreen` is viable as you can still snap (Win + arrows) the fullscreen frame.

**Comment by Stanzilla** [comment]
> @theonlyfoxy, thanks for checking, we are making progress. if you are on the latest WSLg and Windows 11 (all update applied), please create below file with below contents, then window can be snapped by Win+Left or Right key, thanks!
> 
> C:\Users\[Your User Name]\.wslgconfig
> 
> ```
> [system-distro-env]
> WSLG_USE_MSTSC=true
> WESTON_RDP_WINDOW_SNAP_ARRANGE=true
> WESTON_RDPRAIL_SHELL_LOCAL_MOVE=true
> ```

Any updates on this? Is this still needed? What about disabling the white window chrome?

**Comment by undrivendev** [comment]
Any news?

**Comment by meminuygur** [comment]
any news?

**Comment by keli5** [comment]
+1 this is still a very annoying issue 2+ years later

**Comment by sameert89** [comment]
+1 this is the only thing stopping me from liking my setup.

**Comment by moisei** [comment]
@davidhunter22 Enjoy!

```
sudo apt install -y xdotool
xeyes&
xprop -id $(xdotool search --name xeyes) -f _MOTIF_WM_HINTS 32c -set _MOTIF_WM_HINTS "2, 0, 0, 0, 0"
```

**Comment by cyraid** [comment]
I think the whole 'remote desktop' approach isn't entirely too smooth IMO. You're not seeing a native solution, you're seeing a Remote Desktop Protocol in action, think using Remote Desktop into someone's machine.

It 'kinda' works, 'cause you can do things like gaming in the cloud with such techniques, but you'll always have that latency and I don't think video approach is the best. Realistically the best approach would be a native Linux Display Driver. But of course Microsoft may be worried about Kernel boundaries, and running things in Kernel land. I wonder if Microsoft is planning such, and the current RDP is a placeholder? Would like to know thoughts of Microsoft developers  on this one.

**Comment by onomatopellan** [comment]
@cyraid I recommend you to watch this video https://www.youtube.com/watch?v=EkNBsBx501Q and read https://devblogs.microsoft.com/commandline/wslg-architecture/
The future of WSLg is version 2.x, the VAIL+vGPU implementation that uses vRAM instead of system RAM and will bring zero copy, minimal latency.

**Comment by sweetwong** [comment]
I've perfectly solved this problem!!

Bash Script: 
```bash
#!/bin/bash

# Check if any arguments were passed
if [ $# -eq 0 ]; then
  echo "Usage: $0 <window_name>"
  exit 1
fi

# Use the passed argument as the window name
window_name="$1"

# Find all windows matching the window name
window_ids=$(xdotool search --name "$window_name")

# Check if any windows were found
if [ -z "$window_ids" ]; then
  echo "No windows found with the name '$window_name'"
  exit 1
fi

# Loop through all found window IDs and perform the operations
for window_id in $window_ids; do
  echo "Processing window ID: $window_id"
  # Remove window borders
  xprop -id $window_id -f _MOTIF_WM_HINTS 32c -set _MOTIF_WM_HINTS "2, 0, 0, 0, 0"
  # Set window to fullscreen
  wmctrl -ir $window_id -b add,fullscreen
done

echo "All windows have been processed"
```

Use:
```bash
# bash script name: fullscreen, for example: fullscreen xeyes
fullscreen YOUR_WINDOW_NAME
```

![image](https://github.com/user-attachments/assets/35518412-e122-4141-b6ef-865a42abe113)

**Comment by meminuygur** [comment]
> sweetwong

I tried it on phpstorm bot did not work.

**Comment by sameert89** [comment]
> @davidhunter22 , @sameert89 Enjoy!
> 
> ```
> sudo apt install -y xdotool
> xeyes&
> xprop -id $(xdotool search --name xeyes) -f _MOTIF_WM_HINTS 32c -set _MOTIF_WM_HINTS "2, 0, 0, 0, 0"
> ```

Legend! How do you even figure this out? Thanks a bunch!

**Comment by moisei** [comment]
@sameert89 chatgpt

**Comment by sweetwong** [comment]
> > sweetwong
> 
> I tried it on phpstorm bot did not work.

@meminuygur When running this command, ensure that PhpStorm does not launch in a maximized state.

**Comment by antonio24073** [comment]
Hi, 
I'm using [windowgrid ](http://windowgrid.net/) and I need the click sensibility in the title bar.
Please how could I temporally return to old window decoration?
Best regards.

**Comment by MichelTenVoorde** [comment]
Awesome @moisei! Now we just need a way to make IntelliJ (& friends) render its own buttons again.

**Comment by PoProstuWitold** [comment]
Is there any official progress?

**Comment by DorianNaaji** [comment]
### For IntelliJ 
- Add `-Dawt.toolkit.name=WLToolkit` to your VM Options and you'll have clean borders. This is still experimental.
- See https://youtrack.jetbrains.com/issue/JBR-6223/jetbrains-products-look-ugly-in-wslg#focus=Comments-27-9814372.0-0

### For GitKraken 
Change you gitkraken.sh with something like this :
```shell
REAL_PATH="$(readlink -f $0)"
GK_PATH="$(dirname "$(dirname "$(dirname "$REAL_PATH")")")"
CLI="$GK_PATH/resources/app.asar/src/main/static/cli.js"
GITKRAKEN="$GK_PATH/gitkraken"

# Set Wayland environment variables
export XDG_SESSION_TYPE=wayland
export GDK_BACKEND=wayland
export QT_QPA_PLATFORM=wayland
export ELECTRON_OZONE_PLATFORM_HINT=wayland

ELECTRON_RUN_AS_NODE=1 "$GITKRAKEN" "$CLI" "$@"
exit $?
```
That works fine for me.

### For some other apps
https://donfiguerres.com/posts/wsl2/dark-theme-in-wsl2/


I'm still struggling doing it with chrome and cypress for instance though, the last one I provided won't help.

**Comment by niceEli** [comment]
bump, i wish we could use native decorations but nah i guess

**Comment by is1900s** [comment]
> ### For IntelliJ
> * Add `-Dawt.toolkit.name=WLToolkit` to your VM Options and you'll have clean borders. This is still experimental.
> * See https://youtrack.jetbrains.com/issue/JBR-6223/jetbrains-products-look-ugly-in-wslg#focus=Comments-27-9814372.0-0
> 
> ### For GitKraken
> Change you gitkraken.sh with something like this :
> 
> REAL_PATH="$(readlink -f $0)"
> GK_PATH="$(dirname "$(dirname "$(dirname "$REAL_PATH")")")"
> CLI="$GK_PATH/resources/app.asar/src/main/static/cli.js"
> GITKRAKEN="$GK_PATH/gitkraken"
> 
> # Set Wayland environment variables
> export XDG_SESSION_TYPE=wayland
> export GDK_BACKEND=wayland
> export QT_QPA_PLATFORM=wayland
> export ELECTRON_OZONE_PLATFORM_HINT=wayland
> 
> ELECTRON_RUN_AS_NODE=1 "$GITKRAKEN" "$CLI" "$@"
> exit $?
> That works fine for me.
> 
> ### For Chrome
> https://www.reddit.com/r/Fedora/comments/rkzp78/comment/hyvicuy/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button
> 
> ### For some other apps
> https://donfiguerres.com/posts/wsl2/dark-theme-in-wsl2/

GREAT!
<img width="935" height="421" alt="Image" src="https://github.com/user-attachments/assets/fe314eed-8d54-4af3-9da7-aee9adb3a6af" />

**Comment by DorianNaaji** [comment]
Please note that IntelliJ 2026.1 Beta has now native Wayland support, so you won't get ugly borders with newer versions this year. 

You may need the latest WSLg version (`wsl --update`)

**Comment by dermoth** [comment]
Why can't WSLg not draw the borders? I don't think it's even related to the apps themselves, when I run the same apps under VcXsrv I don't get borders (the toolbar is there an looks a bit ugly but it's much less of an annoyance than the borders, especially on a dark theme).

**Comment by fcstr** [comment]
@Copilot fix this

**Comment by ma5d** [comment]
> > ### 对于IntelliJ
> > 
> > * 在虚拟机选项中添加，边界就会干净利落。这仍处于实验阶段。`-Dawt.toolkit.name=WLToolkit`
> > * 参见 https://youtrack.jetbrains.com/issue/JBR-6223/jetbrains-products-look-ugly-in-wslg#focus=Comments-27-9814372.0-0
> > 
> > ### 对于GitKraken
> > 用这样的方式改变你的 gitkraken.sh：
> > REAL_PATH=“$（readlink -f $0）” GK_PATH=“$（dirname ”$（dirname “$（dirname ”$REAL_PATH“）”）“）” CLI=“$GK_PATH/resources/app.asar/src/main/static/cli.js” GITKRAKEN=“$GK_PATH/gitkraken”
> > # 设置Wayland环境变量
> > export XDG_SESSION_TYPE=Wayland export GDK_BACKEND=Wayland export QT_QPA_PLATFORM=Wayland export ELECTRON_OZONE_PLATFORM_HINT=Wayland
> > ELECTRON_RUN_AS_NODE=1 “$GITKRAKEN” “$CLI” “$@” exit $？这对我来说很有效。
> > ### 对于铬
> > https://www.reddit.com/r/Fedora/comments/rkzp78/comment/hyvicuy/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button
> > ### 对于其他一些应用
> > https://donfiguerres.com/posts/wsl2/dark-theme-in-wsl2/
> 
> 太好了！ <img alt="图片" width="935" height="421" src="https://private-user-images.githubusercontent.com/209244698/567580087-fe314eed-8d54-4af3-9da7-aee9adb3a6af.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODA1ODk5NjQsIm5iZiI6MTc4MDU4OTY2NCwicGF0aCI6Ii8yMDkyNDQ2OTgvNTY3NTgwMDg3LWZlMzE0ZWVkLThkNTQtNGFmMy05ZGE3LWFlZTlhZGIzYTZhZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwNjA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDYwNFQxNjE0MjRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05ZGE0MTEwZTFkZjdmYjUzOTcxNjA5NzU4NGYwZGNlZTY2YzA0YzNhMzU2ZGE0OWExNTlmYjFmYmYyMDliOTJmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ.qU7UXGf2s7cXM_1Lb1lp7IMPY9-VxfPVdHZmlGcaNfQ">

你这里有阴影，需要禁用阴影，然后我想问的是，wayland 模式下，idea 不支持中文输入法吗？
